frontend/components/RunArea.js

```diff
@@ -1,11 +1,43 @@
 import { in_textarea_or_input } from "../common/KeyboardShortcuts.js"
-import { html, useEffect, useMemo, useState } from "../imports/Preact.js"
+import { PlutoContext } from "../common/PlutoContext.js"
+import { html, useContext, useEffect, useMemo, useState } from "../imports/Preact.js"
 
-export const RunArea = ({ runtime, onClick, running }) => {
+const upstream_of = (a_cell_id, notebook) => Object.values(notebook?.cell_dependencies?.[a_cell_id]?.upstream_cells_map || {}).flatMap((x) => x)
+
+const all_upstreams_of = (a_cell_id, notebook) => {
+    const upstreams = upstream_of(a_cell_id, notebook)
+    if (upstreams.length === 0) return []
+    return [...upstreams, ...upstreams.flatMap((v) => all_upstreams_of(v, notebook))]
+}
+const hasBarrier = (a_cell_id, notebook) => {
+    return notebook?.cell_inputs?.[a_cell_id]?.is_running_disabled
+}
+
+export const RunArea = ({ runtime, onClick, running, disable, cell_id }) => {
     const localTimeRunning = 10e5 * useMillisSinceTruthy(running)
+    const pluto_actions = useContext(PlutoContext)
     return html`
         <pluto-runarea>
-            <button onClick=${onClick} class="runcell" title="Run"><span></span></button>
+            <button
+                onClick=${() => {
+                    if (!disable) return onClick()
+                    const notebook = pluto_actions.get_notebook() || {}
+                    const barrier_cell_id = all_upstreams_of(cell_id, notebook).find((c) => hasBarrier(c, notebook))
+                    barrier_cell_id &&
+                        window.dispatchEvent(
+                            new CustomEvent("cell_focus", {
+                                detail: {
+                                    cell_id: barrier_cell_id,
+                                    line: 1, // 1-based to 0-based index
+                                },
+                            })
+                        )
+                }}
+                class="runcell"
+                title=${disable ? "Please remove the barrier" : "Run"}
```

Comment:
Perhaps "Please enable the cell" would be closer to the "Enable cell" menu item ?

Reply:
done